### PR TITLE
Allow app to control unity grouping depth

### PR DESCRIPTION
**Pull request by hugary1995**
ref #26556

### CODE REVIEW

File: modules/solid_mechanics/solid_mechanics.mk [hunk 1]
```diff
@@ -2,3 +2,6 @@
 ADDITIONAL_NEML2_DIRS += $(APPLICATION_DIR)/test/neml2
 
 include $(APPLICATION_DIR)/contrib/neml2.mk
+
+# Depth of subfolders used to create unity groups
+app_unity_depth = 2
```
Comment (GiudGiud) [inline modules/solid_mechanics/solid_mechanics.mk:7]: is there a convention on caps / no caps? This is like a Makefile variable which is intended to go across files

I know caps does not change that

File: framework/app.mk [hunk 2]
```diff
@@ -82,22 +82,31 @@ $(eval $(call unity_dir_rule, $(unity_src_dir)))
...
+unity_srcsubdirs_nonmaxdepth := $(filter-out $(non_unity_dirs), $(srcsubdirs_nonmaxdepth))
+unity_srcsubdirs_maxdepth := $(filter-out $(non_unity_dirs), $(srcsubdirs_maxdepth))
 non_unity_srcsubdirs := $(filter $(non_unity_dirs), $(allsrcsubdirs))
 
 # This is a biggie
 # Loop over the subdirectories, creating a rule to create the Unity source file
 # for each subdirectory.  To do that we need to create a unique name using the
 # full hierarchy of the path underneath src
-$(foreach srcsubdir,$(unity_srcsubdirs),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
+$(foreach srcsubdir,$(unity_srcsubdirs_nonmaxdepth),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) -maxdepth 1 \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
```
Comment (GiudGiud) [inline framework/app.mk:108]: should that second 1 for maxdepth be $(shell expr $(app_unity_depth) - 1)  ? we are not going to find srcdir at a different depth than the first time right?

File: modules/solid_mechanics/solid_mechanics.mk (hunk 1 above)
Comment (hugary1995) [reply modules/solid_mechanics/solid_mechanics.mk:7]: I was following the convention for `app_non_unity_dirs`. If there's a different convention, we probably should change both.

File: framework/app.mk (hunk 2 above)
Comment (hugary1995) [reply framework/app.mk:108]: `unity_srcsubdirs_nonmaxdepth` is still recursive, but it contains all directories with different depths, up to `maxdepth-1`. `unity_srcsubdirs_maxdepth` contains only directories at depth `maxdepth`.

The objective here is to generate unity files including every header once and _exactly_ once.

The simplest way of achieving that is:
- For each directory in `unity_srcsubdirs_nonmaxdepth`, include all headers directly (without recursion) underneath it, that is `-maxdepth 1`.
- For each directory in `unity_srcsubdirs_maxdepth`, include all headers recursively.

File: modules/solid_mechanics/solid_mechanics.mk (hunk 1 above)
Comment (GiudGiud) [reply modules/solid_mechanics/solid_mechanics.mk:7]: ok seems fine.

**Review by GiudGiud — APPROVED** [review]

**Comment by loganharbour** [comment]
https://civet.inl.gov/job/2443457/

This somehow broke next by decreasing coverage

```
- `scalar_transport` coverage rate 81.31% is less than the required 85.0%
```

and I have no idea how. I tried invalidating all of the recipes that contribute to coverage to no avail, so I think it's a real issue.

**Comment by GiudGiud** [comment]
are you sure it s not my component PR ? I touched the scalar transport module

**Comment by lindsayad** [comment]
It would be pretty nice to move this upstream in the make system. I want to do this for mfem in the framework

**Comment by hugary1995** [comment]
The logic is already in place, right? I guess the problem is that there is no way of changing app_unity_depth for only mfem and not for the entire framework?

I was chatting with @loganharbour the other day about replacing unity with precompiled headers. We can let each folder decide what headers to precompile.

**Pull request merged**